api/payload.go

```diff
@@ -26,40 +26,79 @@ import (
 // console. The proxy can output this data when asked for verbose output.
 //
 //  {
-//   "containerId": "756535dc6e9ab9b560f84c8...",
-//   "ctlSerial": "/tmp/sh.hyper.channel.0.sock",
-//   "ioSerial": "/tmp/sh.hyper.channel.1.sock"
+//    "containerId": "756535dc6e9ab9b560f84c8...",
+//    "ctlSerial": "/tmp/sh.hyper.channel.0.sock",
+//    "ioSerial": "/tmp/sh.hyper.channel.1.sock",
+//    "numIOStreams: 1
 //  }
 type RegisterVM struct {
 	ContainerID string `json:"containerId"`
 	CtlSerial   string `json:"ctlSerial"`
 	IoSerial    string `json:"ioSerial"`
 	Console     string `json:"console,omitempty"`
+	// NumIOStreams asks for a number of I/O tokens. An I/O token
+	// represents the communication between a container process inside
+	// the VM and a shim process outside the VM. This communication
+	// includes I/O streams (stdin, out, err) but also signals, exit
+	// status, ...
+	// The response frame will contain NumIOStreams I/O tokens.
+	NumIOStreams int `json:"numIOStreams,omitempty"`
+}
+
+// IOResponse is the response data in RegisterVMResponse and AttachVMResponse
+// when the client is asking for I/O tokens from the proxy (NumIOStreams > 0).
+type IOResponse struct {
+	// URL is the URL a shim process should connect to in order to initiate
+	// the I/O communication with the process inside the VM
+	URL string
```

Comment:
`RegisterVM` is used at the POD creation time and one can ask for I/O tokens then for the initial containers.

`AttachVM` can be used at exec time (POD already running), and one can ask for the I/O token corresponding to the exec process.

Both cases should behave the same API wise I believe. With exec/`AttachVM`, we need to spawn a shim with a token as well. I don't really want to keep more state than necessary between create/start and exec, so we get the proxy to tell the runtime the URL again in `AttachVM`.

The URL addition was mine to have some dependency injection going instead of hardcoding where the shim needs to connect to. Now, conceptually, I would associate this URL to the POD object: the shim "connects" to the POD and establishes a link to a process identified by a Token. I couldn't think of a practical case where the processes inside the same pod would need different URLs, so I went for the one URL, multiple tokens choice.